staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -71,29 +71,77 @@ type EgressSelection struct {
 
 // Connection provides the configuration for a single egress selection client.
 type Connection struct {
-	// Type is the type of connection used to connect from client to konnectivity server.
-	// Currently supported values are "http-connect" and "direct".
-	Type string
+	// Protocol is the protocol used to connect from client to the konnectivity server.
+	ProxyProtocol ProtocolType
 
-	// httpConnect is the config needed to use http-connect to the konnectivity server.
+	// Transport defines the transport configurations we use to dial to the konnectivity server
 	// +optional
-	HTTPConnect *HTTPConnectConfig
+	Transport *Transport
 }
 
-type HTTPConnectConfig struct {
+// ProtocolType is a set of valid values for Connection.ProtocolType
+type ProtocolType string
+
+// Valid types for ProtocolType for konnectivity server
+const (
+	// Use http-connect to connect to konnectivity server
+	ProtocolHTTPConnect ProtocolType = "http-connect"
+	// Use grpc to connect to konnectivity server
+	ProtocolGRPC ProtocolType = "grpc"
+	// Direct connection
+	ProtocolDirect ProtocolType = "direct"
+)
+
+// Transport defines the transport configurations we use to dial to the konnectivity server
+type Transport struct {
+	// TCP is the TCP configuration for communicating with the konnectivity server via TCP
+	// Field is optional but one of TCP or UDS must be present if Protocol is http-connect or GRPC
+	// +optional
+	TCP *TCPTransport
+
+	// UDS is the UDS configuration for communicating with the konnectivity server via UDS
+	// Field is optional but one of TCP or UDS must be present if Protocol is http-connect or GRPC
+	// +optional
+	UDS *UDSTransport
+}
+
+// TCPTransport provides the information to connect to konnectivity server via TCP
+type TCPTransport struct {
 	// URL is the location of the konnectivity server to connect to.
 	// As an example it might be "https://127.0.0.1:8131"
 	URL string
 
-	// CABundle is the file location of the CA to be used to determine trust with the konnectivity server.
+	// TLSConfig is the config needed to use TLS when connecting to konnectivity server
+	TLSConfig *TLSConfig
+}
+
+// UDSTransport provides the information to connect to konnectivity server via UDS
+type UDSTransport struct {
+	// UDSName is the name of the unix domain socket to connect to konnectivity server
+	UDSName string
+}
+
+// TLSConfig provides the authentication information to connect to konnectivity server
+// Only used with TCPTransport
+type TLSConfig struct {
+	// caBundle is the file location of the CA to be used to determine trust with the konnectivity server.
+	// Must be absent/empty http-connect using the plain http
+	// Must be configured for http-connect using the https protocol
```

Comment:
we can't use system roots?

Reply:
cc: @cheftako

Reply:
Do you mean that if we can use the kube-apiserver root certificates here? We can, but we need to tweak the flags passed to the konnectivity server to let it accept the root cert.

Anyway, the default transport will be uds, which doesn't require the tlsconfig. We keep the tcp transport for flexibility.

Reply:
For our other client configurations (kubeconfig, webhook admission client config, etc), if no CA bundle is specified, we use system trust roots (matching the behavior of golang's tls.Config with a null RootCAs field value)

Reply:
I'd imagine we could tweak the code to support it, but currently we do not accept an empty CA bundle. TLSConfig was a rename from HTTPConnectConfig, and we're just keeping this here for flexibility. The actual network proxy will not be touching this TLSConfig object and we'll be using UDS so no TLS is required. Given these conditions, is supporting system trust roots still something you'd like us to support?

Reply:
> currently we do not accept an empty CA bundle. TLSConfig was a rename from HTTPConnectConfig, and we're just keeping this here for flexibility

since the current treatment is inconsistent with our other TLS configuration handling, making it consistent while in alpha would make sense to me

> the actual network proxy will not be touching this TLSConfig object and we'll be using UDS so no TLS is required

I assume some configurations will be using it... if not we should drop it entirely. If so, is there a strong reason for it to behave differently than our other TLS configurations that accept CA bundles?

Reply:
ack. Added support for an empty CA bundle. More tests will be added after we have finalized the api.